### DISCUSSION TITLE
Can I write local plugins in TypeScript?

### DISCUSSION THREAD

**Discussion by borekb** [facebook/docusaurus#4657] [Q&A] [answered]
Right now, we have a custom local plugin in a simple `.js` file that we load like this in `docusaurus.config.js`:

```js
module.exports = {
  presets: [
    [
      "@docusaurus/preset-classic",
      {
        docs: {
          remarkPlugins: [require("./our-plugin")],
        },
      },
    ],
  ],
};
```

Can our plugin file be written in TypeScript? Would we need to build it manually, e.g. via `tsc`, or does Docusaurus somehow accept TypeScript natively?

[TypeScript is supported for theme components](https://docusaurus.io/docs/typescript-support) but I'm not sure about plugins.

**Comment by slorber** [comment, marked as answer]
Hmm I don't think we support config in TS for now,  but maybe try and report :)


Maybe it would be complicated to achieve (require running Docusaurus with ts-node) or not convenient (require compiling the plugin to js first), not sure?


However you can probably get some autocompletion support with jsdoc annotations

```js 
/** @type {import('@docusaurus/types').Plugin} */
function MyPlugin() {
  return {
    name,
  };
}
```

![image](https://user-images.githubusercontent.com/749374/115569842-e4700f80-a2bd-11eb-8ad1-76e21a79e758.png)

Can be good enough for now?

Also we'll allow passing a function directly for the plugin so you could write more easily the plugin directly in the config file (if that make sense): https://github.com/facebook/docusaurus/pull/4618

**Reply by borekb** [reply]
Good hints, thanks. I didn't realize that the "problem" is in config itself – I thought that the **config** could stay in **JS** and just the **plugin** would be in **TypeScript** but that's probably nonsense – both would have to support TS.

**Reply by slorber** [reply]
yes,the config and plugins are read by nodejs, so we either have to make nodejs understand TS (ie ts-node, or Deno?) or compile that TS to JS before running

**Reply by borekb** [reply]
We write Next.js apps in TypeScript but are used to the fact that `next.config.js` is plain JavaScript, so this is probably a similar situation.

I don't know what the impacts of running `docusaurus.config.js` through ts-node would be – I think that in the `--transpile-only` mode, it is pretty lightweight but I didn't do any testing.

Plugins can get relatively complex so having full TS power would be awesome 😄 💪.

**Reply by slorber** [reply]
you can already write plugins in TS already, it's just that you'll have to compile them first (like we do for docusaurus plugins)
